packages/yew/src/html/component/lifecycle.rs

```diff
@@ -453,6 +520,49 @@ impl RenderRunner {
                 );
             }
 
+            #[cfg(feature = "hydration")]
+            ComponentRenderState::Hydration {
+                ref mut fragment,
+                ref parent,
+                ref node_ref,
+                ref next_sibling,
+                ref root,
+            } => {
+                // We schedule a "first" render to run immediately after hydration,
+                // for the following reason:
+                // 1. Fix NodeRef (first_node and next_sibling)
+                // 2. Switch from fallback UI to children UI for <Suspense /> component (if it is
+                //    not meant to be suspended.).
+                scheduler::push_component_render(
+                    state.comp_id,
+                    Box::new(RenderRunner {
+                        state: self.state.clone(),
+                    }),
+                );
```

Comment:
Maybe I can clarify what I meant with "push a first_rendered lifecycle event when a component with ComponentRenderState::Hydration is rendered". I think this second could push into the `first_render`, instead of the normal `render` queue of the scheduler. This in turn would trigger a `first_rendered` lifecycle *before any messages are recieved*. For example [in this issue](https://github.com/yewstack/yew/issues/1915#issuecomment-862763865), it should be clear that the current behaviour is to always have `rendered` run before any messages, and I don't see a convincing argument to change this for hydration?

If that leads to problems with only the `BaseSuspense` component (why does this internal comp have to go through message handling in the first place instead of just scheduling a rerender directly?), I'd rather fix that than expose all other existing components to this subtle change in behaviour. Touching the DOM in a message handler isn't unheard of, why break all of those?

Reply:
> why does this internal comp have to go through message handling in the first place instead of just scheduling a rerender directly

Because it needs to mutably borrow the component state (currently suspensions are stored in the suspense itself) and there currently isn't a way to do it.

More special handling creates more edge cases.
Not only you need to write more code and more tests to make sure that they work as expected,
it usually means larger code size as well.

If we can utilize an existing system without special handling, we should do so.

> i.e. make them put in extra effort to be SSR ready?

This is actually not possible to happen if a struct component is SSR ready (with the exception of `BaseSuspense`). This note is actually more of a note of cautious.

Struct components are called in order of `create() -> render() -> update()`, If you want SSR to work, you need to make sure that neither `create()` or `render()` triggers additional renders and that usually means it should not issue messages that results in a re-render (layout changes).
For Event Listeners, currently scheduler will drain all runnables before yielding control to browser so it will not be called before the second render happens.

Most components by their nature, assumes DOM access before effects. So developers already have to rewrite their component to cater this behaviour. It requires quite some efforts to isolate DOM access (if any presents in the component) for struct components anyways so this behaviour change is actually not big even if it does happen.

We should fix this behaviour, but probably not in the form of messing with Suspense Component's rendering lifecycle.

Reply:
From the added documentation

> and the component will start receiving messages immediately after the `view()` function is called after the first time.

This is the unexpected part. As you said, it goes `create() -> view() -> rendered() -> update()` but the docs read as if it were possible during hydration to observe `create() -> view() -> update() -> rendered()` in some cases, since the second render is scheduled without priority over messages. I don't see the necessity of this order to be possible at all, and rather than adding vague warnings, I'd want this to follow the usual order